### ISSUE TITLE
[Feature request] Add option to add PDF417 barcode to invoice

### ISSUE THREAD

**Issue by bgradis** [InvoicePlane/InvoicePlane#1017]
## Expected Behavior
I would like to have an option to add a PDF417 barcode to the invoice which is used in Croatian invoices.

PDF417 spec can be found here (in Croatian language): https://www.hub.hr/sites/default/files/inline-files/2DBK_EUR_Uputa_1.pdf

## Screenshots
Example of an invoice with PDF417 barcode:
![image](https://github.com/InvoicePlane/InvoicePlane/assets/11739989/0f2b4641-9a08-4384-9664-9bbe59eec915)

## Steps to Reproduce

1.
2.
3.
4.

**Comment by marneu** [comment]
Refering: PDF417

Unfortunately not maintained but helpful 
https://github.com/ihabunek/pdf417-php

**Comment by nielsdrost7** [comment]
@bgradis thank you for your issue and @marneu thank you for the suggestion.
On the forums we're talking about all those QR possibilities, because, well, everyone invented their own standards
https://community.invoiceplane.com/t/topic/10627

**Comment by Verony-makesIT** [comment]
@bgradis, 
In the screenshot of your invoice, I also see a QR code. 
Is this the the equivalent of this pdf417 barcode?
If so can you take me a screenshot (without resizing) so I can decode the (text?) data. _(Just to see what the content looks like.)_
A QR code is actually easier to create and put on your invoice.

**Comment by bgradis** [comment]
Hmm, I should explain in detail. 

QR codes on Croatian invoices are usually used for checking invoice validity on the tax administration server. QR code contains URL to https://porezna.gov.hr and sends ZKI or JIR code and invoice date. ZKI code is issuer_id and is always created based on the certificate provided by the tax administration.  JIR code is received from the tax administration server as a response to sending ZKI to the tax server and is also required but in case of network issues, you have 48 hours to send all remaining ZKI codes.

All invoices/receipts that have cash or card transactions must have that code as fiscalisation law states. Transactional invoices don't need that. 

Specifications for fiscalisation and QR codes can be found here: https://www.porezna-uprava.hr/HR_Fiskalizacija/Documents/Fiskalizacija-Tehnicka%20specifikacija%20za%20korisnike_v2.0.pdf (page 48)

Example of QR code used for fiscalisation:
![image](https://github.com/InvoicePlane/InvoicePlane/assets/11739989/7db889da-8a2d-4c67-963c-5e09ed847d75)


PDF417 or maybe it would be better to say HUB-3 PDF417. HUB-3 means Croatian Banking Association (Hrvatska Udruga Banaka) and they define payment slip (not sure if correct translation) type which is called HUB-3 and it looks like this:
![image](https://github.com/InvoicePlane/InvoicePlane/assets/11739989/1c577896-6ad0-4f76-be37-83bb616223ba)
You usually receive that from the utility company for paying the bill. You can see that the PDF417 code is on the screenshot of the HUB3 payment slip. This is handy for scanning with mobile banking apps and accountants which have handheld scanners. I think almost all Croatian banks provide some sort of mobile app with some sort of mPayment variant.
Ad of one of the banks:
![image](https://github.com/InvoicePlane/InvoicePlane/assets/11739989/8d7bff4f-3d34-4e1a-8466-0ae6f201ca99)

PDF417 barcode on the HUB3 payment slip contains all data from it in the structure defined in the following document: https://www.hub.hr/sites/default/files/inline-files/2DBK_EUR_Uputa_1.pdf

So basically instead of typing all invoice data, you can scan the code with your mPayment app and the payment order in the app will be prepared with all fields waiting for you to confirm so the transaction can be made.
So for example, HUB3 payment slip with following data structure and arrangement:
![image](https://github.com/InvoicePlane/InvoicePlane/assets/11739989/af6fb50a-29dd-49e8-95ca-7ec8eb83a1e5)

will have the corresponding PDF417 code: 
![image](https://github.com/InvoicePlane/InvoicePlane/assets/11739989/4047f092-f8b2-4b68-b748-7fa98fd56dcf)

**Comment by Verony-makesIT** [comment]
@bgradis,
Thank you for this detailed explanation.
I had hoped for a moment that the QR code could also be useful for your purpose. Not so.
Maybe these links can help you a bit further on your way?
1. https://github.com/mpdf/mpdf/issues/475
2. https://github.com/ihabunek/pdf417-php

**Comment by nielsdrost7** [comment]
Hey @bgradis someone from the mpdf community forked mpdf and made it working for dotmatrix.
See https://github.com/mpdf/mpdf/issues/475#issuecomment-2028461973

**Comment by naui95** [comment]
Moving the feature for a future version.

**Comment by DavorSaric** [comment]
This would be the most welcome :)

**Comment by nielsdrost7** [comment]
Here is a package for creating such a PDF https://github.com/radoid/pdf417-php

**Comment by DavorSaric** [comment]
If you need someone to test the functionality, I can help with that.

**Comment by pekiZG** [comment]
I would also need this.
Beside the 2D code, here in Croatia law defines required and optional parts of the invoice.

Would be also nice to have separate template only for Croatia or any other country.
So I would implement it in a way that while creating customer you select customer language and template.
And then from that selection you make decision how PDF is generated.

What do you think?

**Comment by nielsdrost7** [comment]
@pekiZG can you get started on this feature request as well please?

I've seen someone post an example for a Slovanian QR code: https://community.invoiceplane.com/t/topic/11618/1

> So I would implement it in a way that while creating customer you select customer language and template.

I agree with you. If you can implement it that way?

**Comment by xam-ps** [comment]
I'd rather add a field to the address called "country" and make the invoice template dependent on that. Language could be ambiguous. E.g. a person speaking English in Slovakia still needs to receive a Slovakian invoice.

**Comment by nielsdrost7** [comment]
Combine this:
> select customer language and template.
And then from that selection you make decision how PDF is generated.

And this:
> add a field to the address called "country" and make the invoice template dependent on that.

Into 1 single solution.
Lean on the "country" field that InvoicePlane offers.
I think it's called `client_country`

Make it so that users from Croatia are happy, but also users from Belgium, France, Germany, the UK and the US.

**Comment by pekiZG** [comment]
@nielsdrost7 I can tackle this in the evenings through next week.
But I'm not sure what to implement exactly.
2D barcode is one thing, QR code is another.
Also invoice is different depending if it's B2G, B2B, B2C
there is fiscalisation and also from next year e-račun (translation from croatian to english is e-invoice)
----
If I would tackle this, I would start with 2D barcode.

**Comment by nielsdrost7** [comment]
@pekiZG perfect.
The rest we'll tackle tonight on Discord, ok?

**Comment by pekiZG** [comment]
@nielsdrost7 we are same time zone.
When will you be on Discord?

**Comment by nielsdrost7** [comment]
> When will you be on Discord?

All day every day.

**Comment by pekiZG** [comment]
@bgradis can you explain for what do you need QR-code?
I dont see purpose unless the invoice is being fiscalized.

2D barcode I would use
https://github.com/radoid/pdf417-php
Was already looking at it before.

**Comment by pekiZG** [comment]
Ill start with implementation of 2d barcode.
QR code can be moved to another issue or v2 version.

Ping this evening with first draft.

**Comment by pekiZG** [comment]
So I have a working 2D barcode.

![Image](https://github.com/user-attachments/assets/a9fc7955-1371-4a79-b6a8-3f8560109278)

I still have some testing to do.
Question is how to enable it.

**Comment by pekiZG** [comment]
@bgradis and @DavorSaric can you test it please.

**Comment by pekiZG** [comment]
Nice, you can also join the conversation in pull request.
https://github.com/InvoicePlane/InvoicePlane/pull/1238

Where to place settings :)

**Comment by DavorSaric** [comment]
Hi, sorry for waiting, I was too busy. I have cloned the invoice plane and edited your commits. It is working, tnx! There is a small issue though, invoice number has to have a letter by our laws for example 1-A-2. And when you it gets populated in the payment app it says that field is not correct, validation errors. That is due to a letter in that field, but only numers can be entered. I override that in my temaplates with this: echo preg_replace('/A(.*)\s/','1-', $invoice->invoice_number) so that should be used in pdf generator also.

**Comment by DavorSaric** [comment]
Yeah, I do not know how to fix this because the invoice plane numbering is not proper for Croatia laws. I have set invoice numbering to 1-A-1 but I had to add year at the end so my invoice numbering is 1-A-1 2025. You cannot set without year because in 2026 it will say the number is allready used when you reset the counter. So I added 2025 to invoice numbering. And I have put that preg_replace because in PDF generator because our laws doesn't allow anything else in invoice numbering. So in Invoice plane gui I have:
1-A-1 2025
In pdf invoice which are sent to clients I have 1-A-1.

And due to that numbering PBZ app says remittance_text is invalid (mode i poziv na broj). They are generated as: HR00-1-A-1 2025 but the letter A and space are not allowed so it should be HR00-1-1-2025. Do you have an idea where should I put preg_replace function in the code? 

Maybe in PDF417Barcode.php:
```
   $paymentData .= "$this->remittance_text\n";                                                          
    $paymentData .= "\n";                                                                           
    $paymentData .= "Placanje po racunu $this->remittance_text\n";
```
I have tried adding here but I got errors, I am not a programmer...

**Comment by DavorSaric** [comment]
Ok, fixed it. That is the way but it was empty because I removed the {{{invoice_number}}} inside GUI section of Invoices in Settings under Remittance_Text. I thought it is set allready to {{{invoice_number}}} because it was gray but it is not.

`  $paymentData .= preg_replace('/Z(.*)\s/','1-', $this->remittance_text) . "\n";                         // Poziv na broj primatelja
    $paymentData .= "\n";                                                                                                           // Šifra namjene
    $paymentData .= "Placanje po racunu " . preg_replace('/Z(.*)\s/','1-', $this->remittance_text) . "\n";  // Opis plaćanja
`

**Comment by nielsdrost7** [comment]
It's understandable, that you want to give comments in your own language in the source code, but if it has to end up in InvoicePlane, let's make the comments in the English language

**Comment by DavorSaric** [comment]
Ok. The PDF417 php library doesn't have those fields described so I cannot extract English words but here is the translation of the whole block. @pekiZG did put new ip_lang,php file in English in his commit also.

```
    $paymentData  = "HRVHUB30\n";                                                                          // Header
    $paymentData .= "EUR\n";                                                                               // Currency
    $paymentData .= sprintf('%015d', intval($invoice->invoice_total*100)) . "\n";   // Total amount
    $paymentData .= "$invoice->client_name\n";                                                             // Client name
    $paymentData .= "$invoice->client_address_1\n";                                                        // Client address
    $paymentData .= "$invoice->client_zip $invoice->client_city\n";                                        // Client postal code and city
    $paymentData .= "$this->recipient\n";                                                                  // Recipient
    $paymentData .= "$invoice->user_address_1\n";                                                          // Recipient Address
    $paymentData .= "$invoice->user_zip $invoice->user_city\n";                                            // Rrecipient postal code and city    $paymentData .= "$this->iban\n";                                                                       // IBAN
    $paymentData .= "HR00\n";                                                                              // modulus or MOD - don't know is this correct term in Eng
    $paymentData .= "$this->remittance_text\n";                                                            // Remittance text - this could be a Reference number also
    $paymentData .= "\n";                                                                                  // Purpose code
    $paymentData .= "Remittance text $this->remittance_text\n";                    //Remittance text - this could be a Payment description 

```

**Comment by DavorSaric** [comment]
@pekiZG yeah that's a bug

![Image](https://github.com/user-attachments/assets/f220e387-4db6-4009-a9be-754cb5bbcf4a)

**Comment by pekiZG** [comment]
I can push changes this evening :)
